code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -69,27 +71,62 @@ public Account getAdminAccount() {
     }
 
     @Override
-    public List<Account> searchAccounts(String name) {
-        List<Account> accounts = new ArrayList<>();
-        accounts.addAll(create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
-                        .and(ACCOUNT.NAME.contains(name))
-                        .and(ACCOUNT.KIND.eq(AccountConstants.USER_KIND)
-                                .or(ACCOUNT.KIND.eq(AccountConstants.ADMIN_KIND))))
-                .orderBy(ACCOUNT.ID.asc()).fetch());
-        return accounts;
+    public List<Account> searchUsers(String username) {
+        try {
+            List<Account> accounts = new ArrayList<>();
+            accounts.addAll(create()
+                    .select(ACCOUNT.fields())
+                    .from(ACCOUNT)
+                    .join(CREDENTIAL)
+                    .on(CREDENTIAL.ACCOUNT_ID.eq(ACCOUNT.ID))
+                    .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
+                            .and(CREDENTIAL.STATE.eq(CommonStatesConstants.ACTIVE))
+                            .and(CREDENTIAL.PUBLIC_VALUE.contains(username))
+                            .and(CREDENTIAL.KIND.eq("password")))
+                    .orderBy(ACCOUNT.ID.asc()).fetchInto(AccountRecord.class));
+            return accounts;
+        } catch (InvalidResultException e) {
+            throw new ClientVisibleException(ResponseCodes.CONFLICT, "MultipleSameUsernames");
+        }
     }
 
     @Override
-    public Account getByName(String name) {
-        return create()
-                .selectFrom(ACCOUNT)
-                .where(ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
-                        .and(ACCOUNT.NAME.eq(name))
-                        .and(ACCOUNT.KIND.eq(AccountConstants.USER_KIND)
-                                .or(ACCOUNT.KIND.eq(AccountConstants.ADMIN_KIND))))
-                .orderBy(ACCOUNT.ID.asc()).fetchOneInto(Account.class);
+    public Account getByUsername(String username) {
+        try {
+            return create()
+                    .select(ACCOUNT.fields())
+                    .from(ACCOUNT)
+                    .join(CREDENTIAL)
+                    .on(CREDENTIAL.ACCOUNT_ID.eq(ACCOUNT.ID))
+                    .where(
+                            ACCOUNT.STATE.eq(CommonStatesConstants.ACTIVE)
+                                    .and(CREDENTIAL.STATE.eq(CommonStatesConstants.ACTIVE))
+                                    .and(CREDENTIAL.PUBLIC_VALUE.eq(username)))
+                    .and(CREDENTIAL.KIND.eq("password"))
+                    .fetchOneInto(AccountRecord.class);
+        } catch (InvalidResultException e) {
+            throw new ClientVisibleException(ResponseCodes.CONFLICT, "MultipleOfUsername");
```

Comment:
What happens when it makes the alter and fails?  Does cattle not start up?  Will it continue and work if you then remove the duplicate from the database and start again?

If it does something rational like that, then I say put it as a footnote in the release notes "if you see `<this error>` on upgrade, use `<steps>` to remove duplicates".  The table is used for apikeys today, and only admin has permission to create specific values, and most people outside the company wouldn't know it was possible to begin with.  So there should be no to very few people with actual duplicates in their database.